### PR TITLE
bin: handle TERM/INT signal

### PR DESCRIPTION
**Pull request by BusyJay**
ref: https://github.com/pingcap/tikv/issues/1053

### CODE REVIEW

File: src/bin/tikv-server.rs [hunk 1]
```diff
@@ -589,6 +593,31 @@ fn get_store_path(matches: &Matches, config: &toml::Value) -> String {
...
+        .name("signal handler".to_owned())
+        .spawn(move || {
+            for sig in trap {
+                match sig {
+                    SIGTERM | SIGINT => {
+                        ch.send(Msg::Quit).unwrap();
+                        break;
+                    }
+                    // TODO: handle more signal
+                    _ => debug!("ignore signal: {}", sig),
```
Comment (ngaut) [inline src/bin/tikv-server.rs]: s/debug/warn/

File: src/bin/tikv-server.rs [hunk 2]
```diff
@@ -589,6 +593,31 @@ fn get_store_path(matches: &Matches, config: &toml::Value) -> String {
...
+fn handle_signal(ch: SendCh<Msg>) {
+    use signal::trap::Trap;
+    use nix::sys::signal::{SIGTERM, SIGINT};
+    let trap = Trap::trap(&[SIGTERM, SIGINT]);
+    thread::Builder::new()
+        .name("signal handler".to_owned())
+        .spawn(move || {
+            for sig in trap {
+                match sig {
+                    SIGTERM | SIGINT => {
```
Comment (ngaut) [inline src/bin/tikv-server.rs]: Please add log for signal handling.

File: src/bin/tikv-server.rs [hunk 3]
```diff
@@ -589,6 +595,42 @@ fn get_store_path(matches: &Matches, config: &toml::Value) -> String {
...
+        .unwrap();
+    handle_signal(ch);
+    h.join().unwrap();
+}
+
+#[cfg(unix)]
+fn handle_signal(ch: SendCh<Msg>) {
+    use signal::trap::Trap;
+    use nix::sys::signal::{SIGTERM, SIGINT};
+    let trap = Trap::trap(&[SIGTERM, SIGINT]);
```
Comment (siddontang) [inline src/bin/tikv-server.rs]: it will block until receiving a signal?

Comment (BusyJay) [inline src/bin/tikv-server.rs]: No, the for loop below will block.

**Comment by siddontang** [comment]
LGTM
We can use this in #1042 too.

File: src/bin/tikv-server.rs (hunk 3 above)
Comment (hhkbp2) [inline src/bin/tikv-server.rs]: Couldn't find any doc or example how trap should be used.
Will it rerun this piece of signal catching code if the specified signal happens? Like a callback.

Comment (BusyJay) [inline src/bin/tikv-server.rs]: Docs: https://tailhook.github.io/signal/signal/trap/index.html
Examples: https://github.com/tailhook/signal/tree/master/examples
It use `sigwait` internally, see also http://man7.org/linux/man-pages/man3/sigwait.3.html

**Comment by BusyJay** [comment]
PTAL

**Comment by hhkbp2** [comment]
LGTM

**Pull request merged**